builtin/providers/aws/resource_aws_ami.go

```diff
@@ -281,11 +286,49 @@ func resourceAwsAmiDelete(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
-	d.SetId("")
+	// Verify that the image is actually removed, if not we need to wait for it to be removed
+	if err := resourceAwsAmiWaitForDestroy(d.Id(), client); err != nil {
+		return err
+	}
 
+	// No error, ami was deleted successfully
+	d.SetId("")
 	return nil
 }
 
+func resourceAwsAmiWaitForDestroy(id string, client *ec2.EC2) error {
+	log.Printf("Waiting for AMI %s to be deleted...", id)
+
+	req := &ec2.DescribeImagesInput{
+		ImageIds: []*string{aws.String(id)},
+	}
+
+	polls := 0
+	for {
+		resp, err := client.DescribeImages(req)
+		if err != nil {
+			if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidAMIID.NotFound" {
+				// AMI was successfully deleted, exit
+				return nil
+			}
+			// Unknown error hit fetching AMI
+			return err
+		}
+
+		if len(resp.Images) == 0 {
+			return nil
+		}
+
+		// Entry retry loop
+		polls++
+		if polls > AWSAMIRetryCount {
+			return errors.New("Timeoout reached waiting for AMI to be deleted")
+		}
+		time.Sleep(AWSAMISleepDuration)
+		continue
+	}
+}
+
```

Comment:
Could we use `resource.StateChangeConf` and `WaitForState()` here? Seems to serve the same purpose

- https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_instance.go#L444-L458